Update `to_ising` and `from_ising` to support `SparsePauliOp`

t-imamichi opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

- Adds `opflow` option to `to_ising` to control the output is `SparsePauliOp` or Opflow's operator.
- Allows `from_ising` to receive `quantum_info`'s operators.

This is part of #433.
The legacy minimum eigensolver of Terra supports only opflow (not `SparsePauliOp`).
We will deprecate the legacy minimum eigensover and opflow when they are deprecated in Terra.

### Details and comments

woodsp-ibm commented on qiskit_optimization/translators/ising.py:
```diff
@@ -13,17 +13,20 @@
 
-def to_ising(quad_prog: QuadraticProgram) -> Tuple[OperatorBase, float]:
+def to_ising(
+    quad_prog: QuadraticProgram, quantum_info: bool = False
```
I imagine this `quantum_info` parameter will get removed once only SparsePauliOp exists. Before then, at some point, we need to transition over to True by default where False is supported but deprecated. So for the release where we introduce this we have it still producing PauliSumOp by default, and somewhere I would think have a warning message saying the default will be changed to SparsePauliOp. When we do that passing False in, to produce a PauliSumOp would be deprecated. We probably should state this intent in the release note. Support both for now defaulting to PauliSumOp, the release after default to SparsePauliOp (deprecating PauliSumOp) and the release after removing that support leaving just SparsePauliOp supported with the flag here gone.

t-imamichi replied:
Yes, you are right. I plan to remove opflow by 1) allow SparsePauliOp with this PR though the default is opflow, 2) change the default with SparsePauliOp, 3) remove opflow. It also makes sense that the warning message and release note mention the plan.

t-imamichi replied:
Added a warning if `opflow` is not set explicitly

t-imamichi commented:
I updated the option name from `quantum_info` to `opflow` because I thought it is easier to understand why it will be deprecated when opflow is deprecated in terra.
I also changed the code to display a warning message if `opflow` option is not set explicitly.

woodsp-ibm commented on qiskit_optimization/algorithms/minimum_eigen_optimizer.py:
```diff
@@ -222,7 +222,7 @@ def solve(self, problem: QuadraticProgram) -> MinimumEigenOptimizationResult:
 
         # construct operator and offset
-        operator, offset = problem_.to_ising()
+        operator, offset = problem_.to_ising(opflow=True)
```
Where we use conversion methods internally, like here, and consume the operators ourselves, in this case via the min eigen solver, we could already move over to using SparsePauliOp now couldn't we. Is there any reason not to? Its what we would recommend right that people start the migration away from PauliSumOp (opflow) - ideally any new code would start out that way though of course its not the default yet for compatibility/migration reasons.

t-imamichi replied:
Unfortunately, we cannot move to `SparsePauliOp` here because `MinimumEigenOptimizer` need to accept the legacy minimum eigensolvers that can deal with only opflow.

woodsp-ibm replied:
Ah right, so it has to stay this way, even if the default of the to_ising gets changed, until the support for the legacy algorithms is removed.

woodsp-ibm commented on releasenotes/notes/support-sprase-pauli-op-00d29000d48c93ca.yaml:
```diff
@@ -0,0 +1,10 @@
+    Updates :func:`~.to_ising` and :func:`~.from_ising` to support
+    ``qiskit.quantum_info.SparsePauliOp``.
+    Adds ``opflow`` option to :func:`~.to_ising` to control the output is
+    opflow's operator (if ``True``) or ``SparsePauliOp`` (if ``False``).
```
```suggestion
    opflow's operator (if ``True``) or ``SparsePauliOp`` (if ``False``).
```

woodsp-ibm commented on qiskit_optimization/translators/ising.py:
```diff
@@ -32,6 +36,8 @@ def to_ising(quad_prog: QuadraticProgram) -> Tuple[OperatorBase, float]:
 
     Args:
         quad_prog: The problem to be translated.
+        opflow: The output object is an OpFlow's operator if True.
```
Juts checking on `is an OpFlow's operator`  - the output here is always a PauliSumOp right, may be we can say then `an opflow PauliSumOp`. In practice I think it has to be a PauliSumOp since thats the only opflow operator listed by primitives as supported i.e. `BaseOperator | PauliSumOp`

t-imamichi replied:
I updated the code so that the return object can be always PauliSumOp with `opflow=True`.

woodsp-ibm commented on qiskit_optimization/translators/ising.py:
```diff
@@ -42,6 +48,15 @@ def to_ising(quad_prog: QuadraticProgram) -> Tuple[OperatorBase, float]:
+            "`opflow` option of `to_ising` is not set explicitly. "
+            "It is set True now, but the default value will be changed to False. "
+            "We suggest setting `opflow=True` explicitly like `to_ising(qp, opflow=True)`.",
+            stacklevel=2,
```
I'll comment here but it partly relates to the text in the release note. Here we default now to True so as to be backward compatible, but this default will be changed to False in a future (the next) release. The use of opflow operators will be deprecated at that point and will not be the default - presumably at that time, here we will detect anyone setting True explicitly and warn to change to False, or preferably remove their explicit setting since the opflow parameter will be removed after the deprecation period. We should recommend moving away from using opflow operators, which can be done right now by passing False here.

t-imamichi commented on qiskit_optimization/translators/ising.py:
```diff
@@ -148,6 +172,11 @@ def from_ising(
         QiskitOptimizationError: if any Pauli term has an imaginary coefficient
         NotImplementedError: If the input operator is a ListOp
     """
+    if isinstance(qubit_op, BaseOperator):
```
Good idea. I will add a pending deprecation warning.

t-imamichi replied:
Added a pending deprecation warning.

a-matsuo reviewed (commented):
I reviewed the code changes thoroughly and everything looks good to me!

a-matsuo reviewed (approved):
I reviewed the code changes thoroughly and everything looks good to me!

woodsp-ibm commented on qiskit_optimization/translators/ising.py:
```diff
@@ -131,6 +150,12 @@ def from_ising(
 
+    .. note::
+
+        The ``qubit_op`` argument can currently accept Opflow operators, but it has been
+        superseded by ``SparsePauliOp``.
+        Opflow will be deprecated in a future release and subsequently removed after that.
```
```suggestion
        The ``qubit_op`` argument can currently accept Opflow operators (OperatorBase type), but have been
        superseded by Qiskit Terra quantum info BaseOperators such as  ``SparsePauliOp``.
        Opflow operator support will be deprecated in a future release and subsequently removed after that.
```
Here is accepts all quantum info BaseOperator types right. Most often it will be SparsePauliOp bit its not restricted to this. I will also note that we can make these class references since we have the intersphinx with Terra.

t-imamichi replied:
Thanks. I include it. cc41551

woodsp-ibm commented on qiskit_optimization/translators/ising.py:
```diff
@@ -148,9 +173,24 @@ def from_ising(
+    # opflow
+    if isinstance(qubit_op, OperatorBase):
+        warn(
+            "The `qubit_op` argument can currently accept Opflow operators, but it has been "
+            "superseded by `SparsePauliOp`. "
+            "Opflow will be deprecated in a future release and subsequently removed after that.",
```
I did not alter the text here as per he above suggestion. But if the suggestion seems better then perhaps this should be the same.

t-imamichi replied:
Thanks. I include it. cc41551

woodsp-ibm commented on releasenotes/notes/support-sprase-pauli-op-00d29000d48c93ca.yaml:
```diff
@@ -0,0 +1,13 @@
+---
+features:
+  - |
+    Updated :func:`~.to_ising` to support ``qiskit.quantum_info.SparsePauliOp``.
+    Added ``opflow`` argument to :func:`~.to_ising` to control the output is
+    opflow operator (if ``True``) or ``SparsePauliOp`` (if ``False``).
+    The default value of ``opflow`` argument is currently ``True``, but it will
+    first be changed to ``False`` and then deprecated in future releases.
```
```suggestion
    Updated :func:`~.to_ising` to support optionally returning ``qiskit.quantum_info.SparsePauliOp``.
    This is achieved by a new ``opflow`` argument to :func:`~.to_ising` to control if the output is
    the opflow operator, as has been done in the past,  (if ``True``) or ``SparsePauliOp`` (if ``False``).
    The default value of ``opflow`` argument is currently ``True``, but it will
    first be changed to ``False``. The parameter will be deprecated and removed in a future release.
```

t-imamichi replied:
Thanks. I include it. cc41551

woodsp-ibm commented on releasenotes/notes/support-sprase-pauli-op-00d29000d48c93ca.yaml:
```diff
@@ -0,0 +1,13 @@
+    The default value of ``opflow`` argument is currently ``True``, but it will
+    first be changed to ``False`` and then deprecated in future releases.
+  - |
+    Updated :func:`~.from_ising` to support ``qiskit.quantum_info.SparsePauliOp``.
+    ``qubit_op`` argument can currently accept opflow operators,
+    but it has been superseded by ``SparsePauliOp``.
+    Opflow will be deprecated in a future release and subsequently removed after that.
```
This text is similar again to the suggestion made for the from_ising note. So again is thats ok probably we want the same/similar here too.

t-imamichi replied:
Thanks. I include it. cc41551

t-imamichi commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -1007,13 +1009,19 @@ def substitute_variables(
 
+        Args:
+            opflow: The output object is an OpFlow's operator if True.
+                Otherwise, it is ``SparsePauliOp``. (default: True)
```
Thank you for your idea. I updated it to refer to `to_ising` function.

woodsp-ibm reviewed (approved):
LGTM. I see the release note has a minor spelling mistake in its name `support-sprase-pauli-op` where its `sprase` instead of `sparse` - but since the name never shows up anywhere outside of looking at the files themselves it's of little consequence.

t-imamichi commented:
Good catch. But, it is a bit complicated to rename reno. So, I will merge it without renaming it.
